state-sync/state-sync-v2/data-streaming-service/src/data_stream.rs

```diff
@@ -291,11 +333,13 @@ impl<T: DiemDataClient + Send + Clone + 'static> DataStream<T> {
         &mut self,
         global_data_summary: GlobalDataSummary,
     ) -> Result<(), Error> {
-        // Check if the stream is complete
         if self.stream_progress_tracker.is_stream_complete()
-            && self.stream_end_notification_id.is_none()
+            || self.request_failure_count >= MAX_REQUEST_RETRY
```

Comment:
i wonder if it might make more sense to put the retry logic in each independent client request task? I would then also put the response conversion logic inside the retry block. each task would retry MAX # of times (with backoff and jitter). the pending_response slot would then get fulfilled with an `Err` if we pass the max retry limit. this would also let us preemptively retry tasks further in the queue (though with more potential for wasted work if the queue HEAD is ultimately bad).

Reply:
Aah, that's an interesting optimization and refactoring to make. To be honest, I didn't think that far ahead. 😄  I kind of wanted to keep this retry logic simple and block on one failure at a time (at least for debugging the MVP).

I'll mark this down as an improvement to make in the future. Although, like you mention, it does make me worry a little bit about wasted work if the stream is in a bad state.